README.md
# Google Cloud IoT Device SDK for Embedded C

The Google Cloud IoT Device SDK for Embedded C is an easy-to-port, open-source C library that connects low-end IoT devices to Google Cloud IoT Core.

The Device SDK supports concurrent Pub/Sub traffic on a non-blocking socket implementation that runs on POSIX, RTOS, and no-OS devices. A Board Support Package (BSP) facilitates portability and provides platform-specific implementations through a set of functions and directories. The BSP helps implement crypto and time functions, networking, Transport Layer Security (TLS), memory management, and random number generation—without working through MQTT internals.

For more details, see the user guide in the `docs` directory.

## Source
To get the source, clone from the `master` branch of the [Google Cloud IoT
Device SDK for Embedded C GitHub
repository](https://github.com/GoogleCloudPlatform/iot-device-sdk-embedded-c.git):

```
git clone https://github.com/GoogleCloudPlatform/iot-device-sdk-embedded-c.git --recurse-submodules
```

### Directory structure

- `bin`: Executables and libraries produced by a build.
- [`doc`](doc): Documentation: Doxygen references, user guide, and porting guide.
- [`examples`](examples): Example source with makefiles. After you build with `make`, this directory will also contain the example executables.
- [`include`](include): Header files of the Device SDK API. You must add this directory to the header include path when compiling your application against the source.
- [`include/bsp`](include/bsp): Header files of the Board Support Package (BSP). Functions declared here must be defined in a device-specific BSP implementation. When compiling your BSP source, make sure this directory is on the include path.
- [`make`](make): Build system configuration files.
- `obj`: Object files generated during a build.
- [`res`](res): Resource files (for example, trusted root CA certificates.)
- [`src`](src): The source files of the Device SDK. A BSP implementation is provided for POSIX.
- [`third_party`](third_party): Third-party open-source components.
- [`tools`](tools): Scripts used by the maintainers of the repository.

## Building

The default build tool for the Device SDK is `make`. It invokes [GCC](https://www.gnu.org/software/gcc/) to produce a native build on a POSIX host. This serves as the default development environment on Ubuntu.

Run the following command to build the Device SDK:

```
make
```

The source can be cross-compiled to other target platforms via custom toolchains and the makefile, or with a specific device SDK’s IDE. For more information, see [the porting guide](doc/porting_guide.md).

### Building a TLS static library

By default, the Device SDK is built with secure connection support and thus requires a third-party TLS implementation to link against. When executing `make`, the build system defaults to mbedTLS.

As a result, the `make` and `make IOTC_BSP_TLS=mbedtls` commands are synonymous.

You can also configure the `make` system to build a library that uses wolfSSL.

```
make clean
make IOTC_BSP_TLS=wolfssl
```

The value of `IOTC_BSP_TLS` determines which script is run.
- mbedTLS: [`res/tls/build_mbedtls.sh`](res/tls/build_mbedtls.sh)
- wolfSSL: [`res/tls/build_wolfssl.sh`](res/tls/build_wolfssl.sh)

The mbedTLS build script includes a git clone and branch checkout of the mbedTLS source (upon confirmation of the license agreement). However, the wolfSSL build script requires you to clone the repository yourself. When you run `make IOTC_BSP_TLS=wolfSSL`, instructions are provided for cloning the repository.

For more details on running the scripts, see [Security](#security).

### Building and executing tests

To build and execute all tests:
1. Run `git submodule init` from the project's root directory to initialize all test dependencies.
1. Run `git submodule update` from the project's root directory to clone all submodules.
1. Run `make tests`.

By default, test execution is the final step of the `tests` build process. You can also execute the tests manually.

```
cd bin/{host_os}/tests
./iotc_utests
./iotc_gtests
./iotc_itests
```

### Building the examples

Before building the examples, build both the Device SDK static library and a TLS library, as described in the preceding sections. Then, complete the steps below to run the examples.

1. Create a project, registry and device in Cloud IoT Core.
2. Create [Cloud IoT Core device credentials](https://cloud.google.com/iot/docs/how-tos/credentials/keys).
3. Follow the steps in the examples README.md files to provision the device credentials and build the client applications.
4. Run `make` in the [`examples`](examples) folder. The `make` process automatically downloads the Google Root CA PEM file to the example directories. The file enables TLS when communicating with Cloud IoT Core.

To securely connect to Cloud IoT Core, a root CA `.pem` file must be in the current working directory of the example executables. By default, the file is in `res/trusted_RootCA_certs/roots.pem` and contains two certificates that validate Cloud IoT Core credentials. The `make` process automatically moves this file from `res/trusted_RootCA_certs/roots.pem` to the correct location.

To maintain a secure connection with Cloud IoT Core, we strongly recommend that you perform frequent security-related firmware updates.

### Cross-compilation

Follow the steps below to perform the cross-compilation process.

- Extend the build system with a new cross-compilation preset in the file `make/mt-config/mt-presets.mk`.
- Create a set of files that represent the Board Support Package (BSP) implementation you've written for your platform. Store these in the directory `src/bsp/platform/TARGET_PLATFORM`.
- Build a TLS BSP implementation to invoke the TLS SDK for your platform. Store these in a new directory `src/bsp/tls/TARGET_TLS_SOLUTION`.
- Build a Cryptography BSP implementation to handle key signatures of JWTs on your target platform in the directory `src/bsp/crypto/TARGET_CRYPTO_SOLUCTION`.

For more details on the cross-compilation process, see [the porting guide](doc/porting_guide.md).

## Security

The Device SDK supports secure connection through a third-party TLS library. The Device SDK is tested against [mbedTLS](https://tls.mbed.org) and [wolfSSL](https://www.wolfssl.com).

This repository does not directly include TLS libraries; you can clone the TLS git repositories and place them in the `third_party/tls/mbedtls` and `third_party/tls/wolfssl` directories, respectively. Running `make` without any parameters will start a build that includes git checkout, build configuration, and compilation of the mbedTLS library.

The Device SDK supports other TLS libraries through the BSP TLS API. For information about configuring the build process to work with your preferred library, see the porting guide in `doc/porting_guide.md`.  Additionally, check the user guide `doc/user_guide.md` to make sure your TLS implementation meets the security requirements to connect to Cloud IoT Core.

## Stability and QA

19 combinations of compilers and feature sets are continuously built. 58 functional, 23 integration and 199 unit tests are executed after each build. Tests are executed against the TLS libraries [mbedTLS](https://tls.mbed.org) and [wolfSSL](https://www.wolfssl.com).

Branch      | Build status
------------|-------------
master      | [![Build Status][travis-badge]][travis-url]

[travis-badge]: https://travis-ci.com/GoogleCloudPlatform/iot-device-sdk-embedded-c.svg?branch=master
[travis-url]: https://travis-ci.com/GoogleCloudPlatform/iot-device-sdk-embedded-c

## Contributing

For information about contributing to this repository, see
[`CONTRIBUTING.md`](CONTRIBUTING.md).

## Learn more

Review the following documentation:

- [`doc/user_guide.md`](doc/user_guide.md): User guide that covers Device SDK features and usage.
- [`doc/porting_guide.md`](doc/porting_guide.md): Porting guide that provides information about porting the Device SDK to target devices.
- Device SDK [API](https://googlecloudplatform.github.io/iot-device-sdk-embedded-c/api/html/) and [BSP](https://googlecloudplatform.github.io/iot-device-sdk-embedded-c/bsp/html/) references.

## License

Copyright 2018-2020 Google LLC

Licensed under the BSD 3-Clause license.

For more information, see [`LICENSE.md`](LICENSE.md).


RELEASE-NOTES.md
# Google Cloud IoT Core Device SDK for Embedded C version  1.0.3
#### April 28, 2021

- Updated the POSIX BSP to use `open` instead of `fopen` for more consistent error detection, [#100](https://github.com/GoogleCloudPlatform/iot-device-sdk-embedded-c/pull/100) by [@pnfisher](https://github.com/pnfisher).
- Fixed a non null-terminated issue in message.topic with ‘iotc_user_subscription_callback_’, [#111](https://github.com/GoogleCloudPlatform/iot-device-sdk-embedded-c/issues/111) by [@csobrinho](https://github.com/csobrinho).
- Added overflow protection to ‘__iotc_calloc()’ calls, [#119](https://github.com/GoogleCloudPlatform/iot-device-sdk-embedded-c/pull/119)
- Updates to the Zephyr Troubleshooting guide, [#104](https://github.com/GoogleCloudPlatform/iot-device-sdk-embedded-c/pull/104) by [@galz10](https://github.com/galz10).
- Fixes to links within the User Guide, [#102](https://github.com/GoogleCloudPlatform/iot-device-sdk-embedded-c/pull/102) by [@mikevoyt](https://github.com/mikevoyt).
- Thank you to all of our contributors!

# Google Cloud IoT Core Device SDK for Embedded C version  1.0.2
#### February 4, 2020

- Wildcard topic name support in MQTT subscriptions, [#89](https://github.com/GoogleCloudPlatform/iot-device-sdk-embedded-c/pull/89) by [@sheindel](https://github.com/sheindel), thank you!
- Documentation overhaul, see Device SDK [API](https://googlecloudplatform.github.io/iot-device-sdk-embedded-c/api/html/) and [BSP](https://googlecloudplatform.github.io/iot-device-sdk-embedded-c/bsp/html/) references.
- Zephyr integration fixes, using Zephyr v1.14.0.
- Improvements on the 'make' build system (wolfssl fixes, addressing rebuild
  issues).

# Google Cloud IoT Core Device SDK for Embedded C version  1.0.1
#### April 8, 2019

- The TLS Board Support Package (BSP) reference implementation for mbedTLS has been updated to use deterministic ECDSA signatures.

- `iotc_get_state_string` may now be used to query error strings for application debugging. See `include/iotc_error.h` to enable this feature.

- Updated the gtest framework compilation standard to c++14.  This does not affect the IoT Device SDK library sources.

- Documentation:
  - Updates to the Device SDK [API](https://googlecloudplatform.github.io/iot-device-sdk-embedded-c/api/html/) and [BSP](https://googlecloudplatform.github.io/iot-device-sdk-embedded-c/bsp/html/) doxygen documentation.
  - [`README.md`](README.md) formatting updates.
  - [`CONTRIBUTING.md`](CONTRIBUTING.md) includes a new pull request processes and a link to the Google style guidelines.

- Networking BSP:
  - Added UDP and IPv6 support.
  - Removed the function `iotc_bsp_io_net_create_socket`. Sockets should now be created in BSP implementations of `iotc_bsp_iot_net_socket_connect`.
  - See `include/bsp/iotc_bsp_io_net.h` for more information about these changes.
  - Removed `src/bsp/platform/posix/iotc_bsp_hton.h` as it was no longer required.

- Time API:
  - The `iotc_time_t` typedef is now defined as an `int64_t`.
  - Added the function `iotc_bsp_time_getmonotonictime_milliseconds`.
  - See `include/bsp/iotc_time.h` for more information about these changes.

# Google Cloud IoT Core Device SDK for Embedded C version  1.0
#### February 22, 2019

- Added FreeRTOS example. See `examples/freertos_linux/Linux_gcc_gcp_iot/README.md` for more information.

- Added Zephyr example.  See `examples/zephyr_native_posix/README.md` for more information.

- Added ATECC608 Secure Element reference implementation for generating JWT signatures with private keys.  See `src/bsp/crypto/cryptoauthlib/iotc_bsp_crypto_cryptoauthlib.c` for more information.

- Connection API Changes:
  - JWT creation and formation was added to a new header file `include/iotc_jwt.h`.
  - `iotc_connect` and `iotc_connect_to` now take the standard MQTT connect credentials of username, password, and client_id.  A JWT, formatted according to [Cloud IoT Core specifications](https://cloud.google.com/iot/docs/how-tos/credentials/jwts), should be passed as the MQTT password when connecting to Cloud IoT Core.
  - For more information, see the Device SDK [API](https://googlecloudplatform.github.io/iot-device-sdk-embedded-c/api/html/index.html) or [BSP](https://googlecloudplatform.github.io/iot-device-sdk-embedded-c/bsp/html/index.html) reference.

# Google Cloud IoT Core Device SDK for Embedded C version 0.7
#### October 1, 2018

The initial release of the Cloud IoT Core Device SDK for Embedded C.  The SDK connects embedded devices to Google Cloud IoT Core via MQTT.

The SDK uses a Board Support Package (BSP) architecture to compartmentalize device-specific code.  This modular approach aids in the porting process. Engineers only need to customize a few files to port the Device SDK to their device. A reference BSP implementation is provided for POSIX systems and two reference TLS implementations are provided for use with either the mbedTLS or wolfSSL embedded libraries. TLS implementations are required for both TLS connectivity and for private-key JWT signatures. JWTs serve as client authentication tokens when connecting to Google Clout IoT Core.

For more documentation on the process of porting the SDK to your embedded device, or for using another TLS library, please see the README.md and the /doc/porting_guide.md documents. For more information about Google Cloud IoT Core, please visit: https://cloud.google.com/iot-core.

The Device SDK is developed on Ubuntu Linux with GNU Make.


include/iotc_error.h
/* Copyright 2018-2020 Google LLC
 *
 * This is part of the Google Cloud IoT Device SDK for Embedded C.
 * It is licensed under the BSD 3-Clause license; you may not use this file
 * except in compliance with the License.
 *
 * You may obtain a copy of the License at:
 *  https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#ifndef __IOTC_ERROR_H__
#define __IOTC_ERROR_H__

#ifdef __cplusplus
extern "C" {
#endif

/*! \file
 * @brief Defines state messages and their numeric codes.
 */

/**
 * @typedef iotc_state_t
 * @brief The state of the client application.
 */
typedef enum {
  /** The SDK function succeeded. @internal Numeric code: 0 @endinternal */ IOTC_STATE_OK = 0,
  /** A timeout occurred. @internal Numeric code: 1 @endinternal */ IOTC_STATE_TIMEOUT,
  /** @cond Internal. Numeric code: 2 */ IOTC_STATE_WANT_READ, /** @endcond */
  /** @cond Internal. Numeric code: 3 */ IOTC_STATE_WANT_WRITE, /** @endcond */
  /** @cond Internal. Numeric code: 4 */ IOTC_STATE_WRITTEN, /** @endcond */
  /** @cond Internal. Numeric code: 5 */ IOTC_STATE_FAILED_WRITING, /** @endcond */
  /** The backoff was applied. @internal Numeric code: 6 @endinternal */ IOTC_BACKOFF_TERMINAL,
  /** The device doesn't have enough memory to fulfull the request. @internal Numeric code: 7 @endinternal */ IOTC_OUT_OF_MEMORY,
  /** @cond Internal. Numeric code: 8 */ IOTC_SOCKET_INITIALIZATION_ERROR, /** @endcond */
  /** @cond Internal. Numeric code: 9 */ IOTC_SOCKET_GETHOSTBYNAME_ERROR, /** @endcond */
  /** @cond Internal. Numeric code: 10 */ IOTC_SOCKET_GETSOCKOPT_ERROR, /** @endcond */
  /** @cond Internal. Numeric code: 11 */ IOTC_SOCKET_ERROR, /** @endcond */
  /** @cond Internal. Numeric code: 12 */ IOTC_SOCKET_CONNECTION_ERROR, /** @endcond */
  /** @cond Internal. Numeric code: 13 */ IOTC_SOCKET_SHUTDOWN_ERROR, /** @endcond */
  /** @cond Internal. Numeric code: 14 */ IOTC_SOCKET_WRITE_ERROR, /** @endcond */
  /** @cond Internal. Numeric code: 15 */ IOTC_SOCKET_READ_ERROR, /** @endcond */
  /** The device isn't connected to an MQTT broker. @internal Numeric code: 16 @endinternal */ IOTC_SOCKET_NO_ACTIVE_CONNECTION_ERROR,
  /** @cond Internal. Numeric code: 17 */ IOTC_CONNECTION_RESET_BY_PEER_ERROR, /** @endcond */
  /** @cond Internal. Numeric code: 18 */ IOTC_FD_HANDLER_NOT_FOUND, /** @endcond */
  /** @cond Internal. Numeric code: 19 */ IOTC_TLS_INITALIZATION_ERROR, /** @endcond */
  /** @cond Internal. Numeric code: 20 */ IOTC_TLS_FAILED_LOADING_CERTIFICATE, /** @endcond */
  /** @cond Internal. Numeric code: 21 */ IOTC_TLS_CONNECT_ERROR, /** @endcond */
  /** @cond Internal. Numeric code: 22 */ IOTC_TLS_WRITE_ERROR, /** @endcond */
  /** @cond Internal. Numeric code: 23 */ IOTC_TLS_READ_ERROR, /** @endcond */
  /** @cond Internal. Numeric code: 24 */ IOTC_MQTT_SERIALIZER_ERROR, /** @endcond */
  /** @cond Internal. Numeric code: 25 */ IOTC_MQTT_PARSER_ERROR, /** @endcond */
  /** @cond Internal. Numeric code: 26 */ IOTC_MQTT_UNKNOWN_MESSAGE_ID, /** @endcond */
  /** @cond Internal. Numeric code: 27 */ IOTC_MQTT_LOGIC_UNKNOWN_TASK_ID, /** @endcond */
  /** @cond Internal. Numeric code: 28 */ IOTC_MQTT_LOGIC_WRONG_SCENARIO_TYPE, /** @endcond */
  /** @cond Internal. Numeric code: 29 */ IOTC_MQTT_LOGIC_WRONG_MESSAGE_RECEIVED, /** @endcond */
  /** @cond Internal. Numeric code: 30 */ IOTC_MQTT_UNACCEPTABLE_PROTOCOL_VERSION, /** @endcond */
  /** @cond Internal. Numeric code: 31 */ IOTC_MQTT_IDENTIFIER_REJECTED, /** @endcond */
  /** The MQTT server is not available. @internal Numeric code: 32 @endinternal */ IOTC_MQTT_SERVER_UNAVAILIBLE,
  /** The MQTT username or password is not correct. @internal Numeric code: 33 @endinternal */ IOTC_MQTT_BAD_USERNAME_OR_PASSWORD,
  /** @cond Internal. Numeric code: 34 */ IOTC_MQTT_NOT_AUTHORIZED, /** @endcond */
  /** @cond Internal. Numeric code: 35 */ IOTC_MQTT_CONNECT_UNKNOWN_RETURN_CODE, /** @endcond */
  /** @cond Internal. Numeric code: 36 */ IOTC_MQTT_MESSAGE_CLASS_UNKNOWN_ERROR, /** @endcond */
  /** @cond Internal. Numeric code: 37 */ IOTC_MQTT_PAYLOAD_SIZE_TOO_LARGE, /** @endcond */
  /** Can't subscribe to the MQTT topic. @internal Numeric code: 38 @endinternal */ IOTC_MQTT_SUBSCRIPTION_FAILED,
  /** @cond Internal. Numeric code: 39 */ IOTC_MQTT_SUBSCRIPTION_SUCCESSFULL, /** @endcond */
  /** Something went wrong. @internal Numeric code: 40 @endinternal */ IOTC_INTERNAL_ERROR,
  /** The SDK is not initialized. Run iotc_initialize() and try again. @internal Numeric code: 41 @endinternal */ IOTC_NOT_INITIALIZED,
  /** @cond Internal. Numeric code: 42 */ IOTC_FAILED_INITIALIZATION, /** @endcond */
  /** @cond Internal. Numeric code: 43 */ IOTC_ALREADY_INITIALIZED, /** @endcond */
  /** A parameter is missing or invalid. @internal Numeric code: 44 @endinternal */ IOTC_INVALID_PARAMETER,
  /** @cond Internal. Numeric code: 45 */ IOTC_UNSET_HANDLER_ERROR, /** @endcond */
  /** Can't create a JWT because the cryptography library in the Board Support Package isn't ported to your platform. To port the cryptography library, customize the <a href="../../bsp/html/d6/d01/iotc__bsp__crypto_8h.html">hardware-specific drivers and routines</a> for your device and then <a href="../../../porting_guide#bsp-code-porting-process">link the library to the client application</a>. @internal Numeric code: 46 @endinternal */ IOTC_NOT_IMPLEMENTED,
  /** @cond Internal. Numeric code: 47 */ IOTC_ELEMENT_NOT_FOUND, /** @endcond */
  /** @cond Internal. Numeric code: 48 */ IOTC_SERIALIZATION_ERROR, /** @endcond */
  /** @cond Internal. Numeric code: 49 */ IOTC_TRUNCATION_WARNING, /** @endcond */
  /** @cond Internal. Numeric code: 50 */ IOTC_BUFFER_OVERFLOW, /** @endcond */
  /** @cond Internal. Numeric code: 51 */ IOTC_THREAD_ERROR, /** @endcond */
  /** @cond Internal. Numeric code: 52 */ IOTC_NULL_CONTEXT, /** @endcond */
  /** @cond Internal. Numeric code: 53 */ IOTC_NULL_WILL_TOPIC, /** @endcond */
  /** @cond Internal. Numeric code: 54 */ IOTC_NULL_WILL_MESSAGE, /** @endcond */
  /** @cond Internal. Numeric code: 55 */ IOTC_NO_MORE_RESOURCE_AVAILABLE, /** @endcond */
  /** @cond Internal. Numeric code: 56 */ IOTC_FS_RESOURCE_NOT_AVAILABLE, /** @endcond */
  /** @cond Internal. Numeric code: 57 */ IOTC_FS_ERROR, /** @endcond */
  /** The memory limiter module isn't installed. To enable the memory limiter, set the <a href="../../../porting_guide.md#development-flags"><code>memory_limiter</code></a> flag in a <a href="../../../porting_guide.md#config">CONFIG</a> argument. @internal Numeric code: 58 @endinternal */ IOTC_NOT_SUPPORTED,
  /** The event processor stopped because the client application {@link iotc_events_stop() shut down the event engine}. @internal Numeric code: 59 @endinternal */ IOTC_EVENT_PROCESS_STOPPED,
  /** @cond Internal. Numeric code: 60 */ IOTC_STATE_RESEND, /** @endcond */
  /** @cond Internal. Numeric code: 61 */ IOTC_NULL_HOST, /** @endcond */
  /** @cond Internal. Numeric code: 62 */ IOTC_TLS_FAILED_CERT_ERROR, /** @endcond */
  /** @cond Internal. Numeric code: 63 */ IOTC_FS_OPEN_ERROR, /** @endcond */
  /** @cond Internal. Numeric code: 64 */ IOTC_FS_OPEN_READ_ONLY, /** @endcond */
  /** @cond Internal. Numeric code: 65 */ IOTC_FS_READ_ERROR, /** @endcond */
  /** @cond Internal. Numeric code: 66 */ IOTC_FS_WRITE_ERROR, /** @endcond */
  /** @cond Internal. Numeric code: 67 */ IOTC_FS_CLOSE_ERROR, /** @endcond */
  /** @cond Internal. Numeric code: 68 */ IOTC_FS_REMOVE_ERROR, /** @endcond */
  /** @cond Internal. Numeric code: 69 */ IOTC_NULL_PROJECT_ID_ERROR, /** @endcond */
  /** The algorithm of the private key for the JWT signature is not supported. @internal Numeric code: 70 @endinternal */ IOTC_ALG_NOT_SUPPORTED_ERROR,
  /** @cond Numeric code: 71 */ IOTC_JWT_FORMATTION_ERROR, /** @endcond */
  /** @cond Numeric code: 72 */ IOTC_JWT_PROJECTID_TOO_LONG_ERROR, /** @endcond */
  /** @cond Numeric code: 73 */ IOTC_NULL_DEVICE_PATH_ERROR, /** @endcond */
  /** The buffer is too small for the data. @internal Numeric code: 74 @endinternal */ IOTC_BUFFER_TOO_SMALL_ERROR,
  /** The buffer for storing formatted and signed JWTs is null. @internal Numeric code: 75 @endinternal */ IOTC_NULL_KEY_DATA_ERROR,
  /** @cond Numeric code: 76 */ IOTC_NULL_CLIENT_ID_ERROR, /** @endcond */

  /** @cond */ IOTC_ERROR_COUNT /** @endcond */ /* Add errors above this line; this should always be last line. */
} iotc_state_t;

/**
 * @brief A flag to translate between {@link ::iotc_state_t state messages} to their numeric codes.
 */
#define IOTC_OPT_NO_ERROR_STRINGS 1

/**
 * @brief Gets the {@link ::iotc_state_t state message} associated with a numeric code.
 *
 * @param [in] e The numeric code of the state message.
 */
extern const char* iotc_get_state_string(iotc_state_t e);

#ifdef __cplusplus
}
#endif

#endif /* __IOTC_ERR_H__ */
